src/xrpld/app/tx/detail/InvariantCheck.cpp

```diff
@@ -1120,4 +1122,46 @@ ValidMPTIssuance::finalize(
         mptokensCreated_ == 0 && mptokensDeleted_ == 0;
 }
 
+//------------------------------------------------------------------------------
+
+void
+ValidPermissionedDomain::visitEntry(
+    bool,
+    std::shared_ptr<SLE const> const& before,
+    std::shared_ptr<SLE const> const& after)
+{
+    if (after->getType() != ltPERMISSIONED_DOMAIN)
+        return;
+    credentialsSize_ = after->getFieldArray(sfAcceptedCredentials).size();
+}
+
+bool
+ValidPermissionedDomain::finalize(
+    STTx const& tx,
+    TER const result,
+    XRPAmount const,
+    ReadView const& view,
+    beast::Journal const& j)
+{
+    if (tx.getTxnType() != ttPERMISSIONED_DOMAIN_SET || result != tesSUCCESS)
+        return true;
+
+    if (!credentialsSize_)
+    {
+        JLOG(j.fatal()) << "Invariant failed: permissioned domain with "
+                           "no rules.";
+        return false;
+    }
+
+    if (credentialsSize_ > PermissionedDomainSet::PD_ARRAY_MAX)
+    {
+        JLOG(j.fatal()) << "Invariant failed: permissioned domain bad "
+                           "credentials size "
+                        << credentialsSize_;
+        return false;
+    }
+
```

Comment:
how about also checking if the accepted credentials are sorted and unique ?

Reply:
added